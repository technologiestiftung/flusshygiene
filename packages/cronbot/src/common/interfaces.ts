// DB

export interface IUser {
  id: number;
  email: string;
}
export interface ISpot {
  id: number;
  name: string;
}

export interface IBasis {
  id: string;
  user: IUser;
  spot: ISpot;
}

export interface IEndpoints extends IBasis {
  measurementsUrl?: string;
  globalIrradianceUrl?: string;
  dischargesUrl?: string;
  discharges: IMeasurement[];
  measurments: IMeasurementConc[];
  globalIrradiances: IMeasurement[];
}

export interface IGeneric extends IBasis {
  url?: string;
  data: IMeasurement[];
}

export interface IError {
  id: string;
  message: string;
  stack?: string;
  source: IObject;
}

// end DB
export interface IObject {
  [key: string]: any;
}

export interface IMeasurement {
  date: string;
  value: number;
}
export interface IMeasurementConc {
  date: string;
  conc_ec: number;
  conc_ie: number;
}
export interface IApiResponse {
  data: any[];
  success: boolean;
  apiVersion: string;
}

// export interface IUser {
//   id: number;
//   email: string;
//   spots: ISpot[];
// }

export interface IUserData extends IUser {
  spots: ISpotData[];
}

// export interface ISpot {
//   id: number;
//   name: string;
//   apiEndpoints: IApiEndpoints;
//   purificationPlants: IPurificationPlant[];
//   genericInputs: IGenericInput[];
// }
export interface ISpotData extends ISpot {
  apiEndpointsData: IApiEndpointsData;
  purificationPlantsData: IPurificationPlantData[];
  genericInputsData: IGenericInputData[];
}
export interface IApiEndpoints {
  measurementsUrl?: string;
  globalIrradianceUrl?: string;
  dischargesUrl?: string;
}
export interface IApiEndpointsData extends IApiEndpoints {
  [key: string]: any;
  measurementsData: IMeasurementConc[];
  globalIrradianceData?: IMeasurement[];
  dischargesData?: IMeasurement[];
}

export interface IPurificationPlant {
  id: number;
  url: string;
  name: string;
}
export interface IPurificationPlantData extends IPurificationPlant {
  data: IMeasurement[];
}
export interface IGenericInput {
  id: number;
  url: string;
  name: string;
}

export interface IGenericInputData extends IGenericInput {
  data: IMeasurement[];
}
export interface IUserCollection {
  users: IUser[];
}
export interface IDataCollection {
  users: IUserData[];
}
