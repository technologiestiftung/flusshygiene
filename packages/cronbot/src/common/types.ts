export type GenericType = "genericInputs" | "purificationPlants";
export type GenericTypeDE = "Generische Werte" | "Kl√§rwerk";
export type ReportType = "admin" | "dataparse" | "dataget" | "ok";
