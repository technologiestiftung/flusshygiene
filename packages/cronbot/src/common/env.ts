// process
const API_HOST: string | undefined = process.env.API_HOST;
const API_VERSION: string | undefined = process.env.API_VERSION;
const AUTH0_AUDIENCE: string | undefined = process.env.AUTH0_AUDIENCE;
const AUTH0_CLIENT_ID: string | undefined = process.env.AUTH0_CLIENT_ID;
const AUTH0_CLIENT_SECRET: string | undefined = process.env.AUTH0_CLIENT_SECRET;
const AUTH0_TOKEN_ISSUER: string | undefined = process.env.AUTH0_TOKEN_ISSUER;
const MAILGUN_DOMAIN: string | undefined = process.env.MAILGUN_DOMAIN;
const MAILGUN_APIKEY: string | undefined = process.env.MAILGUN_APIKEY;
const MAILGUN_FROM: string | undefined = process.env.MAILGUN_FROM;
const MAILGUN_TO: string | undefined = process.env.MAILGUN_TO;
const FLSSHYGN_PREDICT_URL: string | undefined =
  process.env.FLSSHYGN_PREDICT_URL;
// dereived
const API_URL: string | undefined = `${API_HOST}/api/${API_VERSION}`;
let CRONBOT_API_TOKEN: string | undefined = undefined;

export const setApiToken: (token: string) => void = (token) => {
  CRONBOT_API_TOKEN = token;
};

export const getApiToken: () => string | undefined = () => {
  return CRONBOT_API_TOKEN;
};

// #########
if (MAILGUN_DOMAIN === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("MAILGUN_DOMAIN is not defined");
}
if (MAILGUN_APIKEY === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("MAILGUN_APIKEY is not defined");
}
if (MAILGUN_FROM === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("MAILGUN_FROM is not defined");
}
if (MAILGUN_TO === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("MAILGUN_TO is not defined");
}
if (AUTH0_AUDIENCE === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("AUTH0_AUDIENCE is not defined");
}
if (AUTH0_CLIENT_ID === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("AUTH0_CLIENT_ID is not defined");
}
if (AUTH0_CLIENT_SECRET === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("AUTH0_CLIENT_SECRET is not defined");
}
if (FLSSHYGN_PREDICT_URL === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("FLSSHYGN_PREDICT_URL is not defined");
}
if (AUTH0_TOKEN_ISSUER === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("AUTH0_TOKEN_ISSUER is not defined");
}
if (API_HOST === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("API_HOST is not defined");
}

if (API_VERSION === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("API_VERSION is not defined");
}
if (API_URL === undefined && process.env.NODE_ENV !== "test") {
  throw new Error("API_URL cis not defined");
}

if (/undefined/gi.exec(API_URL) !== null) {
  throw new Error("API_URL contains the word undefined");
}

export {
  API_URL,
  API_VERSION,
  API_HOST,
  AUTH0_CLIENT_ID,
  AUTH0_CLIENT_SECRET,
  AUTH0_TOKEN_ISSUER,
  AUTH0_AUDIENCE,
  CRONBOT_API_TOKEN,
  MAILGUN_DOMAIN,
  MAILGUN_APIKEY,
  MAILGUN_FROM,
  MAILGUN_TO,
  FLSSHYGN_PREDICT_URL,
};
