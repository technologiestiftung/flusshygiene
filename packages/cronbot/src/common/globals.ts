import { IGlobals } from "./interfaces";

export const globals: IGlobals = {
  PREDICT: undefined,
  setPredict: function(value) {
    this.PREDICT = value;
  },
  ADMIN_ONLY: undefined,
  setAdminOnly: function(value) {
    this.ADMIN_ONLY = value;
  },
  VERBOSE: undefined,
  setVerbose: function(value) {
    this.VERBOSE = value;
  },
  SKIP_MAIL: undefined,
  setSkipMail: function(value) {
    this.SKIP_MAIL = value;
  },
};
