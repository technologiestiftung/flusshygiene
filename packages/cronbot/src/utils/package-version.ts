import { resolve } from "path";
import fs from "fs";
const packagejson = fs.readFileSync(
  resolve(__dirname, "../../package.json"),
  "utf8",
);
const pkg = JSON.parse(packagejson);
export const version = pkg.version;
