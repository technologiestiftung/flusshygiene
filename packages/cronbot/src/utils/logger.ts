import { globals } from "../common/globals";

export const logger = {
  error: (...args: any[]) => {
    if (globals.VERBOSE === true) {
      console.error("ERROR", ...args);
    }
  },
  warn: (...args: any[]) => {
    if (globals.VERBOSE === true) {
      console.warn("WARNING", ...args); // eslint-disable-line
    }
  },
  info: (...args: any[]) => {
    if (globals.VERBOSE === true) {
      console.info("INFO", ...args); // eslint-disable-line
    }
  },
};
