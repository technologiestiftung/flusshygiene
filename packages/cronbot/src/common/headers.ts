import { version } from "../utils/package-version";
interface Headers {
  [key: string]: string;
}
export const defaultHeaders = {
  "user-agent": `cronbot ${version}`,
};

export function buildHeaders(overrides?: Headers): Headers {
  return { ...defaultHeaders, ...overrides };
}
