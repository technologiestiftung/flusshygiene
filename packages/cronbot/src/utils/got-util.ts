import { API_URL, getApiToken } from "../common/env";
import { buildHeaders } from "../common/headers";

export interface IGotOptionsOverrides {
  url?: string;
  headers?: {
    authorization?: string;
  };
  json?: any;
}
export const gotOptionsFactory: (
  overrides?: IGotOptionsOverrides,
) => Promise<IGotOptionsOverrides> = async (overrides) => {
  try {
    const token = getApiToken();
    // console.log(token);
    if (
      token === undefined &&
      overrides?.headers?.authorization === undefined
    ) {
      throw new Error("gotOptionsFactory could not find a token");
    }

    const headers = buildHeaders({ authorization: token ? token : "" });
    const opts: IGotOptionsOverrides = {
      url: `${API_URL}`,
      headers,
      ...overrides,
    };
    return opts;
  } catch (error) {
    console.error(error);
    throw error;
  }
};
