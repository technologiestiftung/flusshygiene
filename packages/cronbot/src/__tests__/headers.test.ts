import { buildHeaders, defaultHeaders } from "../common/headers";
import { version } from "../utils/package-version";
describe("Testing headers", () => {
  test("should build a header with the default value and some additional ones", () => {
    expect(buildHeaders()).toEqual(defaultHeaders);
    expect(buildHeaders({ auth: "foo" })).toHaveProperty("auth");
    expect(buildHeaders()).toMatchSnapshot({
      "user-agent": `cronbot ${version}`,
    });
    expect(buildHeaders({ "user-agent": "foo" })).toEqual({
      "user-agent": "foo",
    });
  });
});
