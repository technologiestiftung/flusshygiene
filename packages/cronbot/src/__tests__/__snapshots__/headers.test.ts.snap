// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing headers should build a header with the default value and some additional ones 1`] = `
Object {
  "user-agent": "cronbot 2.9.0",
}
`;
